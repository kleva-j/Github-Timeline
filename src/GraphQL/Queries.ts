import { gql } from '@apollo/client';

export const LOAD_JOBS = gql`
  query GetJobs {
    jobs {
      id
      title
      type
      company
      description
      url
      location
      created_at
    }
  }
`;

export const LOAD_SINGLE_JOB = gql`
  query GetJob($id: ID!) {
    job(id: $id) {
      id
      title
      type
      company
      description
      url
      location
      created_at
    }
  }
`;
