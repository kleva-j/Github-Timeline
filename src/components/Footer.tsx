import React from 'react';

interface Props {
  absolute?: boolean;
}

export const Footer: React.FC<Props> = ({ absolute }) => {
  return (
    <>
      <footer
        data-testid="footer"
        className={(absolute ? 'absolute w-full bottom-0 bg-gray-900' : 'relative') + ' pb-6'}
      >
        <div className="container mx-auto px-4">
          <hr className="mb-6 border-b-1 border-gray-700" />
          <div className="flex flex-wrap items-center md:justify-between justify-center">
            <div className="w-full md:w-4/12 px-4">
              <div className="text-sm text-white font-semibold py-1">
                Copyright Â© {new Date().getFullYear()}{' '}
                <a
                  href="http://michaelobasi.com"
                  className="text-white hover:text-gray-400 text-sm font-semibold py-1"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Github Timeline
                </a>
              </div>
            </div>
            <div className="w-full md:w-8/12 px-4">
              <ul className="flex flex-wrap list-none md:justify-end  justify-center">
                <li>
                  <a
                    href="http://michaelobasi.com"
                    className="text-white hover:text-gray-400 text-sm font-semibold block py-1 px-3"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Michael Obasi
                  </a>
                </li>
                <li>
                  <a
                    href="http://michaelobasi.com"
                    className="text-white hover:text-gray-400 text-sm font-semibold block py-1 px-3"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    About Me
                  </a>
                </li>
                <li>
                  <a
                    href="https://frontendtips.netlify.app/"
                    className="text-white hover:text-gray-400 text-sm font-semibold block py-1 px-3"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Blog
                  </a>
                </li>
                <li>
                  <a
                    href="https://github.com/kleva-j/Github-Timeline/blob/develop/LICENCE"
                    className="text-white hover:text-gray-400 text-sm font-semibold block py-1 px-3"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    MIT License
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </footer>
    </>
  );
};
