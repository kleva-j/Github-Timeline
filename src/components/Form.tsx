import React from 'react';

export const SearchForms: React.FC = () => {
  return (
    <form className="flex flex-col sm:flex-row max-w-3xl mx-auto">
      <aside className="flex-1 my-2 md:mx-2 md:ml-0">
        <label htmlFor="desc" className="text-sm font-semibold">Job Description</label>
        <input
          id="desc"
          type="text"
          name="description"
          placeholder="Filter by title, benefits, companies, expertise"
          className="w-full py-2 px-4 border rounded bg-white"
        />
      </aside>

      <aside className="flex-1 my-2 sm:mx-2 md:mr-0">
        <label htmlFor="location" className="text-sm font-semibold">Location</label>
        <input
          type="text"
          id="location"
          name="location"
          placeholder="Filter by city, state, zip code or country"
          className="w-full py-2 px-4 border rounded bg-white"
        />
      </aside>
    </form>
  );
};
