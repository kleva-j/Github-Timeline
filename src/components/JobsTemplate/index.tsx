export * from './template1';
export * from './template2';
