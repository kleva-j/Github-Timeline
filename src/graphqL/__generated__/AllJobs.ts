/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: AllJobs
// ====================================================

export interface AllJobs_jobs_commitment {
  __typename: "Commitment";
  title: string;
  slug: string;
}

export interface AllJobs_jobs_cities_country {
  __typename: "Country";
  name: string;
  isoCode: string | null;
}

export interface AllJobs_jobs_cities {
  __typename: "City";
  name: string;
  country: AllJobs_jobs_cities_country;
}

export interface AllJobs_jobs_countries {
  __typename: "Country";
  name: string;
}

export interface AllJobs_jobs_remotes {
  __typename: "Remote";
  name: string;
  type: string;
  slug: string;
}

export interface AllJobs_jobs_company {
  __typename: "Company";
  name: string;
  slug: string;
  websiteUrl: string;
  logoUrl: string | null;
}

export interface AllJobs_jobs_tags {
  __typename: "Tag";
  name: string;
  slug: string;
}

export interface AllJobs_jobs {
  __typename: "Job";
  id: string;
  title: string;
  slug: string;
  commitment: AllJobs_jobs_commitment;
  cities: AllJobs_jobs_cities[] | null;
  countries: AllJobs_jobs_countries[] | null;
  remotes: AllJobs_jobs_remotes[] | null;
  description: string | null;
  applyUrl: string | null;
  company: AllJobs_jobs_company | null;
  tags: AllJobs_jobs_tags[] | null;
  isPublished: boolean | null;
  isFeatured: boolean | null;
  locationNames: string | null;
  userEmail: string | null;
  postedAt: any;
  createdAt: any;
  updatedAt: any;
}

export interface AllJobs {
  jobs: AllJobs_jobs[];
}
