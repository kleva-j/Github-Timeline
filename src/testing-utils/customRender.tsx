import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';

export const waitForData = () => new Promise((res) => setTimeout(res, 0));

export const customRender = (node: JSX.Element, mocks?: MockedResponse[]) => {
  return {
    ...render(
      <MockedProvider mocks={mocks} addTypename={false}>
        {node}
      </MockedProvider>,
    ),
  };
};
