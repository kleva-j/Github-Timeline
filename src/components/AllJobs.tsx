import React from 'react';
import PropTypes from 'prop-types';
import { ApolloError } from '@apollo/client';

import { RefetchBtn } from 'styled';

import { Job } from './SingleJob';
import { Template2 } from './JobsTemplate';

import loadingGif from 'assets/img/loading-gif.webp';
import errorGif from 'assets/img/error-gif.webp';

interface IProps {
  loading: boolean;
  error: ApolloError | undefined;
  data: {
    jobs: Job[];
  };
  currentPage: number;
  refetch: Function;
}

export const GetAllJobs: React.FC<IProps> = ({ loading, error, data, currentPage, refetch }) => {
  const limit = currentPage * 10;
  const offset = (currentPage - 1) * 10;

  if (loading)
    return (
      <div data-testid="loading" className="text-center my-20 mx-auto w-max">
        <img
          src={loadingGif}
          className="select-none object-contain"
          alt="Loading gif"
        />
      </div>
    );
  if (error)
    return (
      <div data-testid="error" className="text-center my-10">
        <img
          src={errorGif}
          className="select-none object-contain w-max mx-auto"
          alt="Error gif"
        />

        <RefetchBtn onClick={() => refetch()}>Refetch Content</RefetchBtn>
      </div>
    );

  return (
    <article className="w-full" data-testid="jobs-article">
      {data?.jobs.slice(offset, limit).map((item: Job) => (
        <Template2 {...item} key={item.id} />
      ))}
    </article>
  );
};

GetAllJobs.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.any,
  currentPage: PropTypes.number.isRequired,
  refetch: PropTypes.func.isRequired
};
