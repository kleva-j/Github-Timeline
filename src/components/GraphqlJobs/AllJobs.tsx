import { FC, useRef, useEffect, useCallback, MouseEventHandler, MouseEvent, useState } from 'react';
import { v5 as uuidNameSpace, v4 as getNS } from 'uuid';
import { AllJobs } from 'graphqL/__generated__/AllJobs';
import { LOAD_JOBS } from 'graphqL/queries.graphql';
import { useQuery } from '@apollo/client';
import { UseModal } from 'hooks/useModal';
import { Button } from 'styled';

import { SkeletonLoader } from './SkeletonLoader';
import { FilterIcon } from './icons/filter';
import { ErrorComponent } from './Error';
import { JobModal } from './JobModal';
import { Card } from './Card';

import useOnClickOutside from 'hooks/useOnClickOutside';
import styled from 'styled-components';
import tw from 'tailwind.macro';

const GenerateNameSpace: string = getNS();

const config = {
  notifyOnNetworkStatusChange: true,
  context: {
    requestTrackerId: uuidNameSpace('JOBS', GenerateNameSpace),
  },
};

interface IState {
  jobSlug: string;
  companySlug: string;
}

export const GetAllJobs: FC = () => {
  const { loading, error, data, refetch } = useQuery<AllJobs>(LOAD_JOBS, config);

  const { toggle, isVisible } = UseModal();
  
  const modalRef = useRef<HTMLDivElement>(null);
  
  const closeModal = useCallback((): void => toggle(false), [toggle]);
  
  const [state, setState] = useState<IState>({ companySlug: '', jobSlug: '' });

  const handleClick: MouseEventHandler = (event: MouseEvent<HTMLElement>): void => {
    const dataSet = event.currentTarget?.dataset;
    if (dataSet) {
      setState(() => ({
        jobSlug: dataSet.jobslug || '',
        companySlug: dataSet.companyslug || '',
      }));
    }
    toggle();
  };

  useOnClickOutside(modalRef, closeModal);

  useEffect(() => {
    if (isVisible) {
      document.body.style.overflow = 'hidden';
    }
    return () => {
      document.body.style.overflow = 'unset';
    };
  }, [isVisible]);

  return (
    <div className="max-w-screen-lg m-auto">
      <div className="flex flex-col justify-between md:items-center mt-20 md:mt-36 mb-10 md:flex-row">
        <Heading>Find the job that you <br /> really love</Heading>
        <FilterBtn>
          Filter
          <FilterIcon style={{ marginLeft: 8 }} />
        </FilterBtn>
      </div>
      <div className="grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-2 xl:grid-cols-3 gap-8">
        {loading && <>{SkeletonLoader({ count: 6 })}</>}
        {(data?.jobs ?? []).map((item) => (
          <Card {...item} key={item.id} handleClick={handleClick} />
        ))}
        {error && <ErrorComponent refetch={() => refetch()} />}
      </div>
      {isVisible && (
        <JobModal handleClose={closeModal} ref={modalRef} companySlug={state.companySlug} jobSlug={state.jobSlug} />
      )}
    </div>
  );
};

const Heading = styled('h1')`
  ${tw`text-4xl font-bold text-gray-800`}

  font-family: Lato, sans-serif;
`;

const FilterBtn = styled(Button)`
  ${tw`text-gray-700 cursor-pointer`}
`;
