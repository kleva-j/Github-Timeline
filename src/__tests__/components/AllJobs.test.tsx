import { cleanup } from '@testing-library/react';

import { LOAD_JOBS } from '../../GraphQL/Queries';
import { GetAllJobs } from '../../components/AllJobs';
import { customRender, waitForData } from '../../testing-utils/customRender';

const mock = {
  request: {
    query: LOAD_JOBS,
    variables: {},
  },
  result: {
    data: {
      jobs: [
        {
          id: 'XXXX',
          title: 'Buck',
          type: 'bulldog',
          company: 'company',
          description: 'description',
          url: 'some_random_url',
          location: 'location',
          created_at: 'created_at',
        },
      ],
    },
  },
};
const errorMock = {
  request: mock.request,
  error: new Error('An error occurred'),
};

describe('GetAllJobs', () => {
  afterEach(cleanup);

  test('it should test the loading state.', async () => {
    const { getByTestId } = customRender(<GetAllJobs />, [mock]);
    const loading = getByTestId('loading');
    expect(loading.textContent).toContain('Loading...');
  });

  test('it should test the success state.', async () => {
    const { container, getByText, getByTestId } = customRender(<GetAllJobs />, [mock]);
    await waitForData();
    expect(container.firstChild).toMatchSnapshot();
    expect(getByText('Github Jobs Listings')).toBeDefined();
    const title = getByTestId('title');
    const type = getByTestId('type');
    const location = getByTestId('location');
    expect(title.textContent).toContain('Buck');
    expect(type.textContent).toContain('bulldog');
    expect(location.textContent).toContain('location');
  });

  test('it should test the error state.', async () => {
    const { getByTestId } = customRender(<GetAllJobs />, [errorMock]);
    await waitForData();
    const error = getByTestId('error');
    expect(error.textContent).toContain('Oops, Error occured!!!');
  });
});
