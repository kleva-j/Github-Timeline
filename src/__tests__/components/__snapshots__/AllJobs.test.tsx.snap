// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAllJobs it should test the error state. 1`] = `
<div
  class="text-center my-10"
  data-testid="error"
>
  <img
    alt="Error gif"
    class="select-none object-contain w-max mx-auto"
    src="error-gif.webp"
  />
  <button
    class="sc-bdfBwQ sc-fubCfw kOVXAr bsBkXx"
  >
    Refetch Content
  </button>
</div>
`;

exports[`GetAllJobs it should test the loading state. 1`] = `
<div
  class="text-center my-20 mx-auto w-max"
  data-testid="loading"
>
  <img
    alt="Loading gif"
    class="select-none object-contain"
    src="loading-gif.webp"
  />
</div>
`;

exports[`GetAllJobs it should test the success state. 1`] = `
<article
  class="w-full"
  data-testid="jobs-article"
>
  <aside
    class="sc-eCssSg cPLCZE"
    data-testid="job-XXXX"
  >
    <div
      class="flex flex-col sm:flex-row justify-between sm:items-center"
    >
      <span
        class="font-light text-gray-600"
      >
        Mar 12, 2021
      </span>
      <div
        class="flex justify-start py-1"
      >
        <button
          class="sc-iBPRYJ gEACbz text-gray-700 truncate whitespace-nowrap"
          data-testid="type-XXXX"
        >
          bulldog
        </button>
        <button
          class="sc-iBPRYJ gEACbz text-gray-700 truncate whitespace-nowrap"
          data-testid="location-XXXX"
        >
          location
        </button>
      </div>
    </div>
    <div
      class="mt-2"
      data-testid="title-XXXX"
    >
      <a
        class="text-xl text-gray-700 font-bold hover:text-gray-600"
        href="some_random_url"
        rel="noopener noreferrer"
        target="_blank"
      >
        Buck
      </a>
    </div>
    <div
      class="flex flex-col sm:flex-row justify-between sm:items-center mt-4"
    >
      <a
        class="text-blue-600 hover:underline whitespace-nowrap order-last sm:order-first mt-2 sm:mt-0"
        href="some_random_url"
        rel="noopener noreferrer"
        target="_blank"
      >
        Read more
      </a>
      <div>
        <a
          class="flex items-center"
          href="some_company_url"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="avatar"
            class="mx-4 w-10 h-10 object-contain hidden sm:block"
            src="some_company_logo"
          />
          <h1
            class="text-gray-600 font-bold"
          >
            company
          </h1>
        </a>
      </div>
    </div>
  </aside>
</article>
`;
